hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/TestRMFailoverProxyProvider.java

```diff
@@ -0,0 +1,320 @@
+        .getProxy(any(YarnConfiguration.class), any(Class.class),
+        eq(mockAdd1));
+
+    // Mock RMProxy methods to generate different proxy
+    // based on different IP address.
+    when(mockRMProxy.getRMAddress(
+        any(YarnConfiguration.class),
+        any(Class.class))).thenReturn(mockAdd2);
+    when(mockRMProxy.getProxy(
+        any(YarnConfiguration.class),
+        any(Class.class), eq(mockAdd2))).thenReturn(mockProxy2);
+
+    // Perform Failover and get proxy again from failover proxy provider
+    fpp.performFailover(actualProxy1.proxy);
+    FailoverProxyProvider.ProxyInfo <RMProxy> actualProxy2 = fpp.getProxy();
+    assertEquals("ConfiguredRMFailoverProxyProvider " +
+        "doesn't generate expected proxy after failover",
+        mockProxy2, actualProxy2.proxy);
+
+    // check the proxy is different with the one we created before.
+    assertNotEquals("ConfiguredRMFailoverProxyProvider " +
+        "shouldn't generate same proxy after failover",
+        actualProxy1.proxy, actualProxy2.proxy);
+
+    // verify that mockRMProxy.getProxy() has been one with each address
+    verify(mockRMProxy, times(1))
+        .getProxy(any(YarnConfiguration.class), any(Class.class),
+        eq(mockAdd1));
+    verify(mockRMProxy, times(1))
+        .getProxy(any(YarnConfiguration.class), any(Class.class),
+        eq(mockAdd2));
+
+    // Mock RMProxy methods to generate the first proxy again
+    when(mockRMProxy.getRMAddress(
+        any(YarnConfiguration.class),
+        any(Class.class))).thenReturn(mockAdd1);
+    when(mockRMProxy.getProxy(
+        any(YarnConfiguration.class),
+        any(Class.class), eq(mockAdd1))).thenReturn(mockProxy1);
+
+    // Perform Failover and get proxy again from failover proxy provider
+    fpp.performFailover(actualProxy2.proxy);
+    FailoverProxyProvider.ProxyInfo <RMProxy> actualProxy3 = fpp.getProxy();
+
+    // check the proxy is the same as the one we created before.
+    assertEquals("ConfiguredRMFailoverProxyProvider " +
+        "doesn't generate expected proxy after failover",
+        mockProxy1, actualProxy3.proxy);
+
+    // verify that mockRMProxy.getProxy() has still only been invoked twice
+    verify(mockRMProxy, times(1))
+        .getProxy(any(YarnConfiguration.class), any(Class.class),
+        eq(mockAdd1));
+    verify(mockRMProxy, times(1))
+        .getProxy(any(YarnConfiguration.class), any(Class.class),
+        eq(mockAdd2));
+  }
+
+  /**
+   * Test that the {@link AutoRefreshRMFailoverProxyProvider}
+   * will loop through its set of proxies when
+   * and {@link AutoRefreshRMFailoverProxyProvider#performFailover(Object)}
+   * gets called.
+   */
+  @Test
+  public void testAutoRefreshFailoverChange() throws Exception {
+    conf.setClass(YarnConfiguration.CLIENT_FAILOVER_NO_HA_PROXY_PROVIDER,
+        AutoRefreshRMFailoverProxyProvider.class, 
+        RMFailoverProxyProvider.class);
+
+    class MockProxy extends Proxy implements Closeable {
```

Comment:
You may want to do this MockProxy used by the other methods instead of repeating.